### ISSUE TITLE
Prototype a sparse discrete Laplace backend beside the dense reference

### ISSUE THREAD

**Issue by Veronica0206** [Veronica0206/Gtheory4LLM#3]
Dense fitting constructs an `(n*q) x r` random-design matrix and an `r x r` conditional Hessian at every likelihood evaluation. Increasing its size ceilings cannot provide the planned full-panel operating range.

After response/interface extraction, implement a private Matrix-based algebra prototype first. Retain the current whitened model `eta = baseline + W*u`, standard-normal penalty, and first-order Laplace objective. Keep dense preparation, parameterization, and acceptance semantics as references.

Scope:

- Store source incidence and ordering once; construct/update sparse `W` without a dense intermediate. Handle zero and fixed source factors explicitly.
- Build sparse Hessians from existing response curvature, using the full softmax block covariance for categorical outcomes.
- Use factor-object solves and log determinants so fill-reducing permutations cannot corrupt a Newton step. Never densify a factor to extract its determinant.
- Separate `nnz(W)`, symbolic factor fill, measured allocation/RSS, and runtime. A sparse `W` alone is not evidence that the Hessian factor is affordable.
- Keep the prototype internal and opt-in until equivalence and rejection tests pass; no silent fallback or estimator substitution.

Begin with fixed-covariance binary logit/probit and then ordinal. Joint/categorical cases remain on dense fitting until separately qualified, never coerced to independent binary fits. TMB is a candidate for the later AD implementation, not a dependency committed by this issue.

Prototype qualification: prepare frozen fixed-parameter fixtures and tolerances before implementation, then require agreement in predictor, score, Hessian, mode, log determinant, and marginal likelihood. Add a small predeclared set of fitted and intentionally rejected smoke comparisons and report storage reduction. These checks are part of this issue and cannot be postponed until broader qualification.

Blocked by #2 only. This issue does not depend on completing #4. After this prototype passes its bounded checks, #4 performs the comprehensive fitted, rejection, permutation, and equivariance matrix. Passing the prototype checks does not establish full backend qualification. No 21,600-row fit is required here.

Reference: [Matrix sparse Cholesky documentation](https://stat.ethz.ch/R-manual/R-devel/library/Matrix/html/Cholesky-methods.html).

**Comment by Veronica0206** [comment]
## Scope clarification: the 0.2 sparse prototype is binary and ordinal

Recording a narrow wording correction. This changes no numerical requirement in this issue.

One scope bullet above reads as though the sparse Hessian assembly should carry full softmax categorical curvature, while the support paragraph later in the same issue keeps categorical and joint fitting on the dense backend until separately qualified. For 0.2.0 the latter governs:

> The Matrix sparse prototype qualified in #3 targets native binary and ordinal fitting. Categorical and joint sparse fitting remain unsupported and stay on the dense backend until separately qualified; no independent-binary substitution is permitted.

This matches what has actually been built and validated. The merged sparse Hessian assembly refuses categorical curvature outright rather than approximating it, and the refusal names the dense backend; the sparse conditional-mode solver refuses categorical responses before doing any work. The fitted smoke matrix now being frozen covers binary and ordinal only.

Sparse categorical and joint fitting are deferred to a separate post-0.2 issue rather than left as an apparently unfinished bullet here.

**Comment by Veronica0206** [comment]
## Completion evidence

The bounded sparse discrete Laplace prototype is complete. Evidence below, and an explicit list of what is **not** claimed.

### Completed

| | Where |
| --- | --- |
| Sparse `W` built from triplets, never a dense intermediate | `R/discrete_sparse.R`, #24 |
| Sparse conditional Hessian `I + W'CW` | #25 |
| Factor-object solves and log determinant, permutation-invariant | #26 |
| Sparse conditional mode and marginal Laplace value | #27 |
| Fixed-parameter qualification against a reference frozen before any sparse code | `validation-studies/discrete-sparse-reference/`, 1354 quantities |
| Fitted qualification against a contract frozen before any sparse fitted evaluator | `validation-studies/discrete-sparse-fitted-smoke/`, ten cases |
| Accepted / rejected / refused parity, exact | #30, Layer 3 |
| Exact-zero fitted boundary preserved | `zero_variance_parameters` as contract |
| Cross-platform execution | Linux, Windows, macOS, verified per-platform in job logs |
| Storage and factor fill recorded | `nnz(W)`, `nnz(H)`, factor entries, dense-cell comparison |
| Runtime, allocation and peak RSS on an accepted default-envelope panel | `validation-studies/discrete-sparse-profile/`, #31 |

The resource evidence is a 1200-row panel with 112 random coordinates, sitting exactly at `max_observations` and inside `max_random_dimension`, so **both backends run under default limits and no guard was raised to obtain a comparison**. Both accepted, both selecting `primary_tight`:

```
                                   dense        sparse
elapsed                           83.26 s      12.19 s
peak RSS (process)                 707 MB       483 MB
Rprofmem recorded (>= 1 KB)     26,412 MB    11,885 MB
largest single recorded alloc      1.03 MB      0.07 MB
marginal evaluations                  765          765
design storage              134,400 cells    2,400 nnz
```

Both fits used 765 evaluations, so the timing difference is not explained by evaluation count. Optimizer-path identity is **not** asserted — that is #4's work.

No speedup threshold is claimed, and the profile is evidence rather than a performance test. On the 300–400 row smoke panels sparse is *slower*, because those are deliberately sized so dense remains available as the reference.

### One outer refactor, not a second fitter

`.gt_fit_discrete()` gained a private marginal-evaluator seam. Optimizer, restarts, tight validation, stationarity and acceptance remain single-sourced; there is no `.gt_fit_discrete_sparse()`. Dense is unchanged and remains the default and the numerical reference.

### Carried forward, and explicitly not #3 blockers

| | Goes to |
| --- | --- |
| 2400-row cold conditional-solve scale risk | #4 and #8 |
| Comprehensive backend qualification | #4 |
| Warm starts | #5, still blocked by #4 |
| Automatic differentiation | #6, conditional on profiling |
| Public sparse dispatch and backend-scoped limits | later integration issue |
| 21,600-row demonstration | #8 |
| Sparse categorical and joint fitting | unsupported, deferred past 0.2 |

### The scale-risk specimen

A 2400-row engineering profile (`panel_md5 = bad937e27a0ce3ee587b47ca1ae41b25`) was **rejected**. It is retained as characterization and excluded from every performance figure. It does **not** expand or qualify anything claimed here.

At the selected final point the two evaluators agree to 4e-16 in the mode and exactly in the objective, and the final mode is converged. The failure is generated upstream: 40 of 584 evaluations exhausted the 60-iteration inner budget and returned the invalid-evaluation sentinel, putting discontinuities into the outer objective. Replaying those specific points through dense shows dense reaching the same budget at the same points, and in two of eight sampled points returning invalid where sparse still produced a finite value.

The evidence therefore supports a cold conditional-solve limitation that is **not sparse-specific**. Stated that way rather than as a full dense-fit equivalence, because 2400 rows exceeds `max_observations` and the dense guard was not raised to obtain a comparison.

This was discovered before #4's protocol was frozen, so #4 can cover the risk deliberately — but as a post-discovery characterization specimen, never as a predeclared result.

### Support boundary

Native binary and ordinal only. Categorical and joint sparse fitting remain on the dense backend and are refused by the sparse path rather than approximated. No public backend selector ships with this work.

**Comment by Veronica0206** [comment]
## Correction to the completion evidence above

Two figures in the completion comment were mislabelled. Correcting them here rather than editing the record. Neither affects what #3 completed, and no timing, memory or storage figure changes.

**The inner-iteration counts describe valid solves only.** Both mode solvers return `list(valid = FALSE, inner_converged, inner_gradient)` for an evaluation they could not complete — with no `inner_iterations` field. A failed evaluation therefore cannot contribute an iteration count, and the instrumentation cannot attribute its cause.

**The 2,400-row specimen.** The comment said "40 of 584 evaluations exhausted the 60-iteration inner budget". Both halves were wrong. Recounted:

```
632 evaluations   584 valid   48 invalid   40 valid solves ending at iteration 60
```

584 was the valid count, not the total, and the 48 invalid evaluations were absent from the record. Those are the evaluations returning the sentinel that puts discontinuities into the outer objective, and their cause is **not** established to be budget exhaustion.

This makes the characterization sharper rather than weaker: the discontinuity source is larger than reported and less explained than implied. It does not change the classification — at the selected final point the two evaluators still agree to 4e-16 in the mode and exactly in the objective, and the evidence still supports a cold conditional-solve limitation that is not sparse-specific.

**The 1,200-row profile**, likewise: 5 dense and 4 sparse invalid evaluations, against 1 and 3 valid solves ending at iteration 60.

Both corrections, plus a fail-closed fix to the RSS extraction in the reproduction wrapper, are in Veronica0206/Gtheory4LLM#32.

#3 remains complete. The carried-forward items are unchanged, and the sharper 2,400-row figures strengthen rather than alter the risk evidence handed to #4 and #8.
